### ISSUE TITLE
Global SEO Image taking precedence over content SEO images

### ISSUE THREAD

**Issue by aarmitage** [nystudio107/craft-seomatic#1284]
### Describe the bug
Global SEO image is taking precedence over content SEO images and entry specific overrides.

### To reproduce
Steps to reproduce the behaviour:
1. Set a custom image as the SEO image source in Global SEO and upload/select required asset (transform SEO image also selected)
2. Click on Content SEO, select content channel (blog) and in General, set a content section specific SEO Image Source by choosing From Asset Field, Source Field: Main Image.
3. Create an entry in the respective content section (blog)
4. Preview the page and see the global SEO image instead of the content SEO image.

### Expected behaviour
I'd expect to see the Content SEO image, not the Global SEO which is stubbornly showing on each entry in the content section where either a default or override has been specified.

### Screenshots
<img width="1592" alt="general-seo-image-1" src="https://user-images.githubusercontent.com/1152884/217898590-61104f35-9baf-4f42-968c-75dc892f6b5b.png">
<img width="1592" alt="content-seo-2" src="https://user-images.githubusercontent.com/1152884/217898559-428373cf-35dc-4bb9-81b2-13edb0a431e6.png">
<img width="1592" alt="entry-default-3" src="https://user-images.githubusercontent.com/1152884/217898573-a2635c8d-0b65-457c-add2-c4a6ed0be572.png">
<img width="1592" alt="entry-override-4" src="https://user-images.githubusercontent.com/1152884/217898586-d500bb31-47c5-441c-b76c-967c972f74ed.png">
<img width="1592" alt="entry-fields-5" src="https://user-images.githubusercontent.com/1152884/217898581-44e89f23-7678-48b9-a5e0-4a864a85fa5e.png">

### Versions
- Plugin version: 3.4.47
- Craft version: 3.7.63.1

**Comment by khalwat** [comment]
The only time I've seen this, there was some kind of error transforming the image. Do you see any errors logged in your `web.log`?

**Comment by aarmitage** [comment]
Ok, thanks. We've looked in the logs but can't see anything that looks likely here - looked for references to transforms in the file. Searching for SEOmatic returned 355 matches and the ones of these we did parse through appeared unrelated (SEOmatic appears as part of various objects in many 'info' log entries).

**Comment by khalwat** [comment]
Are you able to replicate this issue in any other channel(s) on this same site, or on any other sites that use SEOmatic?

**Comment by aarmitage** [comment]
Frustratingly no, we haven't. We have SEOmatic installed on most of our sites and this is the first time we've seen it. There are other channels on the same site which are working fine. I've compared the settings which are identical. I've tried changing the image selection, changing and resaving settings but no change.

**Comment by khalwat** [comment]
I have tried to reproduce it locally, but have been unable to.

Can you check the db to see if you have multiple `seomatic_metabundles` for that section?

**Comment by aarmitage** [comment]
Ha! Yes there is. They have different IDs, but one has a NULL typeId. Apart from that and dateCreated column they look identical. Which would you suggest is removed?

**Comment by khalwat** [comment]
Interesting. Remove the `null` typeId one. It's not a big deal to remove one or both, you'll just lose whatever settings are there for that particular section, SEOmatic will recreate it for you on demmand.

**Comment by aarmitage** [comment]
Ok thanks - removed them both for good measure and the entry overrides are working now, but the site-wide settings are still taking precedence over the content settings for that section.
On 13 Feb 2023 at 23:05 +0000, Andrew Welch ***@***.***>, wrote:
> Interesting. Remove the null typeId one. It's not a big deal to remove one or both, you'll just lose whatever settings are there for that particular section, SEOmatic will recreate it for you on demmand.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***>

-- 
We provide flexible working for our team around the needs of our clients, 
so you might have received this email outside of your own working hours, 
but we don't expect to hear back from you until it's convenient for you.

If you've received this email by mistake, we're sorry for bothering you. It 
may contain information that's confidential, so please delete it without 
sharing it. And, if you let us know we can try and stop it from happening 
again.

A Digital Ltd is registered in England and Wales. Registered 
Address: Suite 3 Unit 2 Meadowbank Business Park, Shap Road, Kendal, 
Cumbria, LA9 6NY. Company No. 7041426. VAT Registration No. 924 1361 47. 
Prices where quoted are exclusive of VAT.

**Comment by khalwat** [comment]
And you have set up the Content SEO settings for this section, for each entry type for that section?

**Comment by aarmitage** [comment]
Yes, we do - have looked again here internally, and still don't see anything amiss. Each field is populated, the correct field's set for assets etc.

**Comment by khalwat** [comment]
Send me a database dump and your `composer.json` to: `andrew@nystudio107.com` and I can try it locally.

**Comment by khalwat** [comment]
Also something you can try:

What if you change the setting in Content SEO or in the SEO Settings field to "Custom URL" and paste in some random URL to an image?

Does it override and work properly then?

**Comment by aarmitage** [comment]
Yes, that's worked first time in the content SEO section for that channel, on both entry types. It seems to be the reference to the `entry.mainImage` field it's choking on which is a standalone field (so not in a matrix block or anything). Transforms on/off don't seem to make any difference.

I'll get a DB and composer.json over to you. 

Thank you.

**Comment by khalwat** [comment]
...and what if you choose a _different_ image in the `entry.mainImage`? One that works in another Section, perhaps?

**Comment by aarmitage** [comment]
Tried that, no luck. Used an image that is being successfully used by SEOmatic in a different section and applied it to the channel that won't work.

**Comment by khalwat** [comment]
Got it all up and running locally and reproduced the issue — it looks like you’re using Craft’s built-in updater to update plugins, so SEOmatic is locked at 3.4.47 — which had a regression.

Updating it to SEOmatic 3.4.49 (the latest currently) caused me to no longer be able to reproduce this issue locally.

You might consider using Composer & semver for your plugin updates:

https://nystudio107.com/blog/updating-craft-cms-without-headaches#4-embrace-composer

**Comment by thom-moresoda** [comment]
Hi @khalwat  I'm having an identical looking problem with a site running
Craft - 4.3.1
SEOmatic - 4.0.23

It seems to only affect one channel that I've noticed so far

- seomatic_metabundles in DB look correct
- Content SEO settings are set up
- neither Custom URL or Custom Image override the default SEO image
- There were some failed transforms in the queue logs, but I've cleared those & the problem persists

Any help appreciated!

**Comment by thom-moresoda** [comment]
Update!

running `craft clear-caches/compiled-templates` resolves the issue 👍

**Comment by khalwat** [comment]
@thom-moresoda you left a message saying that clearing the caches fixed the issue?

**Comment by thom-moresoda** [comment]
Hi @khalwat I did! Unfortunately on further testing it turned out to be a false positive & the issue was still there

**Comment by khalwat** [comment]
Well, the issue is probably the failed transforms. If Craft fails when transforming the image, the URL will be empty.

**Comment by thom-moresoda** [comment]
Thanks @khalwat - we'll investigate the transforms further

**Comment by thom-moresoda** [comment]
Hi @khalwat we've updated to the latest version of the plugin & the transforms now look to be working - annoyingly the SEO image issue is still persisting.

I think the only thing we've not done at this point is removing the plugin completely & starting fresh - obviously not ideal as there's a fair bit of content in there. Do you have any suggestions we could try?

**Comment by jwb009** [comment]
I just ran across this and solved this same issue. For me the issue occurred when I changed my asset volumes to use a new Filesystem that happened to be a local folder. Initially, I had this filesystem set with `Files in this filesystem have public URLs` _unchecked_. Once I _checked_ that setting, the transforms worked and everything was back to normal. 

I didn't think I wanted my Transforms Filesystem to have public URL's, but I guess it's necessary for SEOmatic.

**Comment by thom-moresoda** [comment]
@jwb009  Great shout, but no luck for me. All the assets already have the public url setting set to true. I did re-save all the asset & filesystem settings just to make sure though!

@khalwat Just to double down on the transforms not being the cause - I went in to the SEOmatic content settings just now & turned off the transform image toggle option, so any custom images shouldn't be getting a transform applied
